File: wagerproof-ios-native/WagerproofKit/Sources/WagerproofServices/CFBTeamsService.swift
import Foundation
import Supabase
import WagerproofModels

/// Hydrates the CFB team reference cache (`cfb_teams`) once per launch. The
/// dry-run slate uses this for AP-rank rows, logos, conferences, and CFBD colors.
public actor CFBTeamsService {
    public static let shared = CFBTeamsService()
    public init() {}

    private var didLoad = false

    private struct TeamRow: Decodable {
        let teamName: String
        let abbr: String?
        let conference: String?
        let classification: String?
        let color: String?
        let altColor: String?
        let logo: String?
        let logoDark: String?

        enum CodingKeys: String, CodingKey {
            case teamName = "team_name"
            case abbr, conference, classification, color, logo
            case altColor = "alt_color"
            case logoDark = "logo_dark"
        }
    }

    public func ensureLoaded() async {
        if didLoad { return }
        let cfb = await CFBSupabase.shared.client
        guard let rows: [TeamRow] = try? await cfb
            .from("cfb_teams")
            .select("team_name, abbr, conference, classification, color, alt_color, logo, logo_dark")
            .execute()
            .value, !rows.isEmpty
        else { return }

        didLoad = true
        let teams = rows.map {
            CFBTeamReference(
                teamName: $0.teamName,
                abbr: $0.abbr,
                conference: $0.conference,
                classification: $0.classification,
                color: $0.color,
                altColor: $0.altColor,
                logo: $0.logo,
                logoDark: $0.logoDark
            )
        }
        await MainActor.run { CFBTeamAssets.install(teams) }
    }
}

File: wagerproof-android-native/core/services/src/main/kotlin/com/wagerproof/core/services/CFBTeamsService.kt
package com.wagerproof.core.services

import com.wagerproof.core.models.CFBTeamAssets
import com.wagerproof.core.models.CFBTeamReference
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Columns
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

/**
 * Hydrates the CFB team reference cache (`cfb_teams`) once per process. The
 * dry-run slate uses this for AP-rank rows, logos, conferences, and CFBD
 * colors. Silent failure — cards degrade to name-derived fallbacks.
 */
object CFBTeamsService {

    private val mutex = Mutex()
    private var didLoad = false

    @Serializable
    private data class TeamRow(
        @SerialName("team_name") val teamName: String,
        val abbr: String? = null,
        val conference: String? = null,
        val classification: String? = null,
        val color: String? = null,
        @SerialName("alt_color") val altColor: String? = null,
        val logo: String? = null,
        @SerialName("logo_dark") val logoDark: String? = null,
    )

    suspend fun ensureLoaded() {
        mutex.withLock {
            if (didLoad) return
            val rows = runCatching {
                SupabaseClients.cfb
                    .from("cfb_teams")
                    .select(
                        columns = Columns.raw(
                            "team_name, abbr, conference, classification, color, alt_color, logo, logo_dark",
                        ),
                    )
                    .decodeList<TeamRow>()
            }.getOrNull() ?: return
            if (rows.isEmpty()) return

            didLoad = true
            CFBTeamAssets.install(
                rows.map {
                    CFBTeamReference(
                        teamName = it.teamName,
                        abbr = it.abbr,
                        conference = it.conference,
                        classification = it.classification,
                        color = it.color,
                        altColor = it.altColor,
                        logo = it.logo,
                        logoDark = it.logoDark,
                    )
                },
            )
        }
    }
}
